ballerina/variable_validator.bal

```diff
@@ -0,0 +1,147 @@
+            }
+        }
+        if fieldNode.getSelections().length() > 0 {
+            parser:Selection[] selections = fieldNode.getSelections();
+            foreach parser:Selection subSelection in selections {
+                self.visitSelection(subSelection, data);
+            }
+        } 
+    }
+
+    public isolated function visitFragment(parser:FragmentNode fragmentNode, anydata data = ()) {
+        foreach parser:Selection selection in fragmentNode.getSelections() {
+            self.visitSelection(selection, data);
+        }
+    }
+
+    public isolated function visitArgument(parser:ArgumentNode argumentNode, anydata data = ()) {
+        map<parser:VariableDefinition> variableDefinitions = <map<parser:VariableDefinition>> data;
+        string variableName = <string> argumentNode.getVariableName();
+        parser:Location location = argumentNode.getValue().location;
+        if variableDefinitions.hasKey(variableName) {
+            parser:VariableDefinition variableDefinition = variableDefinitions.get(variableName);
+            self.visitedVariableDefinitions.push(variableName);
+            if self.variables.hasKey(variableName) {
+                argumentNode.setKind(getArgumentTypeKind(variableDefinition.kind));
+                Scalar|error value= <Scalar> self.variables.get(variableName);
+                if value is error {
+                    string message = string`Variable "$${variableName}" got invalid value ${self.variables.get(variableName).toString()};`;
+                    self.errors.push(getErrorDetailRecord(message, location));
+                } else {
+                    self.setArgumentValue(value, argumentNode, location);
+                }
+            } else if variableDefinition?.defaultValue != () {
+                argumentNode.setKind(getArgumentTypeKind(variableDefinition.kind));
+                Scalar defaultValue = <Scalar> (<parser:ArgumentValue> variableDefinitions.get(variableName)?.defaultValue).value;
+                self.setArgumentValue(defaultValue, argumentNode, location);
+            } else {
+                string message = string`Variable "$${variableName}" of required type ${variableDefinition.kind} was not provided.`;
+                self.errors.push(getErrorDetailRecord(message, location)); 
+            }
+        } else {
+            string message = string`Variable "$${variableName}" is not defined.`;
+            self.errors.push(getErrorDetailRecord(message, location)); 
+        }
+    }
+
+    public isolated function setArgumentValue(Scalar value, parser:ArgumentNode argument, parser:Location location) {
+        if argument.getKind() == parser:T_IDENTIFIER {
+            argument.setValue(value);
+        } else if getTypeNameFromValue(value) == getTypeName(argument) {
+            argument.setValue(value);
+        } else if value is int && getTypeName(argument) == FLOAT {
+            argument.setValue(<float> value);
```

Comment:
Since we are coercing the int value to float here, shall we remove the coercing in the `ValidatorVisitor`?

Reply:
Currently, this function is used only for variable values. Parameter arguments coercing is done at the ```ValidatorVisitor```. parameters which have values, won't reach here.

Reply:
Right, my bad. But, in that case, we do not need to coerce them here, right?